Improve column visibility on Report List pages

nublarsec opened the pull request:
This implements several changes outlined in #86 for the following pages:

* Vulnerability Reports
* Config Audit Reports
* CIS Kube Bench Reports

### Better default column widths

Using `flex-grow` styles to make some nicer default percentages for the columns.

### ~~Show only most relevant info in columns with long values~~

* ~~For the report names, I'm removing "daemonset/replicaset/statefulset" from the beginning of the string before displaying it. Users can still see the full report name in the tooltip or details page.~~
* ~~I'm doing the same for the image references by removing the image registry URL. When reviewing these results I think it's most important for users to see `repo/image:tag`. They can still see the full registry URL in the tooltip or details page.~~

### Collapse severity counts into a single column

We can reclaim a lot of space by collapsing the 4-5 columns of individual counts into a single column. I recently added color badges in #83, and these are easy to scan and understand next to each other. The display is just like the DrawerItem on the details pages.

I've added tooltips with the severity label, so users can still hover over a count to determine if it is a `HIGH` or `MEDIUM`.

### Combine sorting for severity counts

Collapsing severity counts into one column means they can't be individually sorted. However I don't think that's a very likely use case.

Instead I think most users want to sort by "How bad are these?". If we simply do a cascading sort (Critical first, then High, then Medium, etc) I think that accomplishes it perfectly. This way sorting by the Results column shows the "worst" reports at the top.

An example of all these changes together:
<img width="1235" alt="Screen Shot 2022-04-22 at 11 11 56 AM" src="https://user-images.githubusercontent.com/103554953/164776631-aaa49115-8a97-49f1-97f5-6627f04bb5d5.png">

danielpacak commented on src/vulnerabilityreports/page.tsx:
```diff
@@ -115,8 +111,10 @@ export class VulnerabilityReportPage extends React.Component<{ extension: Render
 
 function renderName(name: string) {
+    // drop replicaset/daemonset/statefulset from displayed name; keep in tooltip
+    const shortName = name.substring(name.indexOf('-')+1)
```
I'm not sure we want to do this. At least not is scope of this PR

danielpacak commented on src/vulnerabilityreports/page.tsx:
```diff
@@ -44,21 +42,19 @@ export class ClusterVulnerabilityReportPage extends React.Component<{ extension:
-                    {title: "Medium", className: "medium", sortBy: sortBy.medium},
-                    {title: "Low", className: "low", sortBy: sortBy.low},
-                    {title: "Unknown", sortBy: sortBy.unknown},
+                    {title: "Results", className: "results", sortBy: sortBy.results},
```
In other places counts grouped by severity are called "Summary".  I'm not sure we can make this column sortable. Would is the sort algorithm?

danielpacak replied:
In other places counts grouped by severity are called "Summary".  I'm not sure we can make this column sortable. Would is the sort algorithm?

danielpacak commented on src/vulnerabilityreports/page.tsx:
```diff
@@ -126,19 +124,21 @@ function renderScanner(scanner: Scanner) {
 function renderImage(report: VulnerabilityReport) {
     const imageRef = report.getImageRef()
+    // drop registry from displayed image reference; keep in tooltip
+    const shortImageRef = imageRef.substring(imageRef.indexOf('/')+1)
```
We use this extension also for debugging Starboard so removing parts of names or tooltips might be confusing. Let's propose such changes in issues or dedicated PR.

danielpacak commented on src/vulnerabilityreports/page.tsx:
```diff
@@ -126,19 +124,21 @@ function renderScanner(scanner: Scanner) {
 function renderImage(report: VulnerabilityReport) {
     const imageRef = report.getImageRef()
+    // drop registry from displayed image reference; keep in tooltip
+    const shortImageRef = imageRef.substring(imageRef.indexOf('/')+1)
```
We use this extension also for debugging Starboard so removing parts of names or tooltips might be confusing. Let's propose such changes in issues or dedicated PR.

nublarsec replied:
Removed; will discuss in a separate PR

danielpacak commented on src/ciskubebenchreports/page.tsx:
```diff
@@ -28,29 +25,30 @@ export class CISKubeBenchReportsPage extends React.Component<{ extension: Render
-                    {title: "Warn", className: "pass", sortBy: sortBy.warn},
-                    {title: "Info", className: "xinfo", sortBy: sortBy.info},
-                    {title: "Pass", className: "pass", sortBy: sortBy.pass},
+                    {title: "Results", className: "results", sortBy: sortBy.results},
```
```suggestion
                    {title: "Summary", className: "summary", sortBy: sortBy.results},
```
```suggestion
                    {title: "Results", className: "results", sortBy: sortBy.results},
```

danielpacak replied:
```suggestion
                    {title: "Summary", className: "summary", sortBy: sortBy.results},
```
```suggestion
                    {title: "Results", className: "results", sortBy: sortBy.results},
```

danielpacak commented on src/configauditreports/page.tsx:
```diff
@@ -42,19 +41,17 @@ export class ClusterConfigAuditReportPage extends React.Component<{ extension: R
-                    {title: "High", sortBy: sortBy.high},
-                    {title: "Medium", sortBy: sortBy.medium},
-                    {title: "Low", sortBy: sortBy.low},
+                    {title: "Results", className: "results", sortBy: sortBy.results},
```
Results -> Summary

danielpacak replied:
Results -> Summary

danielpacak reviewed (changes requested):
Nice improvement! I left a few comments. I also noticed that badges are not aligned in the listing. Is is something we can fix as well?

![badge_aligh](https://user-images.githubusercontent.com/1322923/165259129-539723d4-4bf0-48fb-8313-c1992ea0b379.png)

nublarsec commented:
@danielpacak thanks for the feedback; I've made those changes.

### Regarding sorting:

The sorting algorithm is simply to sort by Critical first, then High, then Medium, etc. Here's the code:

```typescript
sortingCallbacks={{
    ...
    [sortBy.summary]: (report: ConfigAuditReport) => [
        report.report.summary.criticalCount,
        report.report.summary.highCount,
        report.report.summary.mediumCount,
        report.report.summary.lowCount
    ],
    ...
}}
```

This doesn't allow for sorting by individual columns, but I think it's a good compromise—users instead can sort by _overall_ severity. If I click the column and sort by Summary descending, I know the "worst" report is at the top.

### Regarding alignment:

It's a good point—the variable length of each count means they get out of horizontal alignment. Let me take a look at the styles for Badges and try some fixed sizing. I still think we're regaining a lot of space by collapsing the columns together (especially important if we're not truncating long names).

nublarsec commented:
Ok, I've made the badges in the Summary columns fixed-width instead of variable. The size still scales with the user's zoom level rather than a static pixel size. The counts are now right-aligned.

Here's an example with an old Nginx that has a double- and triple-digit findings in each severity:
<img width="1155" alt="Screen Shot 2022-04-27 at 1 28 01 PM" src="https://user-images.githubusercontent.com/103554953/165625733-9b2b465c-e5f2-4566-a637-da2d8e1cc6d3.png">

I've made this change on the 3 report lists we have so far:

* Vulnerability Reports
* Config Audit Reports
* CIS Kube Bench Reports

@danielpacak what do you think?

danielpacak reviewed (commented):
This is looking great @nublarsec The code looks good as well. I was just wondering if it's my screen that the text is not aligned with the checkbox in the first column? Do you think we can align it somehow?

![lens-vertical-text-align](https://user-images.githubusercontent.com/1322923/168011235-21a2cfe4-c36c-47f0-9b1b-73e3d1598410.png)

nublarsec commented:
@danielpacak Not just you, it's a good point. The badge size threw off the vertical alignment a bit. I changed the badge size to `small`. Also Flex makes it pretty easy to middle align: I've added this to the styles for the 3 report pages: 

```css
      display: flex;
      align-items: center;
```

Now they look like this:
<img width="1181" alt="Screen Shot 2022-05-12 at 9 49 24 AM" src="https://user-images.githubusercontent.com/103554953/168128238-cc505478-ef01-4144-b085-75f9a8f50421.png">
